src/main/java/dev/openfeature/sdk/LayeredEvaluationContext.java

```diff
@@ -0,0 +1,235 @@
+        }
+        return merged;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return (hookContexts == null || hookContexts.isEmpty())
+                && (invocationContext == null || invocationContext.isEmpty())
+                && (clientContext == null || clientContext.isEmpty())
+                && (transactionContext == null || transactionContext.isEmpty())
+                && (apiContext == null || apiContext.isEmpty());
+    }
+
+    @Override
+    public Set<String> keySet() {
+        return Collections.unmodifiableSet(ensureKeySet());
+    }
+
+    private Set<String> ensureKeySet() {
+        if (this.keySet != null) {
+            return this.keySet;
+        }
+
+        var keys = new HashSet<String>();
+
+        if (hookContexts != null) {
+            for (int i = 0; i < hookContexts.size(); i++) {
+                var current = hookContexts.get(i);
+                keys.addAll(current.keySet());
+            }
+        }
+
+        if (invocationContext != null) {
+            keys.addAll(invocationContext.keySet());
+        }
+        if (clientContext != null) {
+            keys.addAll(clientContext.keySet());
+        }
+        if (transactionContext != null) {
+            keys.addAll(transactionContext.keySet());
+        }
+        if (apiContext != null) {
+            keys.addAll(apiContext.keySet());
+        }
+        this.keySet = keys;
+        return keys;
+    }
+
+    private Value getFromContext(EvaluationContext context, String key) {
+        if (context != null) {
+            return context.getValue(key);
+        }
+        return null;
+    }
+
+    private Value getFromContext(ArrayList<EvaluationContext> context, String key) {
+        if (context == null) {
+            return null;
+        }
+
+        for (int i = context.size() - 1; i >= 0; i--) {
+            var current = context.get(i);
+            var value = getFromContext(current, key);
+            if (value != null) {
+                return value;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Value getValue(String key) {
+        var hookValue = getFromContext(hookContexts, key);
+        if (hookValue != null) {
+            return hookValue;
+        }
+        var invocationValue = getFromContext(invocationContext, key);
+        if (invocationValue != null) {
+            return invocationValue;
+        }
+        var clientValue = getFromContext(clientContext, key);
+        if (clientValue != null) {
+            return clientValue;
+        }
+        var transactionValue = getFromContext(transactionContext, key);
+        if (transactionValue != null) {
+            return transactionValue;
+        }
+        return getFromContext(apiContext, key);
+    }
+
+    @Override
+    public Map<String, Value> asMap() {
+        var keySet = ensureKeySet();
+        var keys = keySet.size();
+        if (keys == 0) {
+            return new HashMap<>(0);
```

Comment:
```suggestion
            return Collections.emptyMap();
```

Same as https://github.com/open-feature/java-sdk/pull/1717/files#r2543987740